ovs-dpdk.md

```diff
@@ -0,0 +1,227 @@
+
+
+To better scale the workloads across cores, Multiple pmd threads can be created and pinned to CPU cores by explicitly specifying pmd-cpu-mask. eg: To spawn 2 pmd threads and pin them to cores 1, 2
+
+run the following commands:
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-socket-mem="1024,0" 
+# ovs-vsctl set Open_vSwitch . other_config:pmd-cpu-mask=6
+
+```
+
+
+Set dpdk-init so OVS would initialize and support DPDK ports:
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-init=true
+```
+ 
+Restart openvswitch service
+```
+# sudo systemctl restart openvswitch
+```
+ 
+Verify DPDK is enabled:
+```
+# ovs-vsctl get Open_vSwitch . iface_types
+[dpdk, dpdkr, dpdkvhostuser, dpdkvhostuserclient, geneve, gre, internal, ipsec_gre, lisp, patch, stt, system, tap, vxlan]
+```
+
+### Make the changes persistent
+
+Create a new service unit in /etc/systemd/system/ovs-dpdk.service with content based on the following:
+
+```
+[Unit]
+Description=ovs-dpdk configuration script
+Before=openvswitch.service
+
+[Service]
+Type=simple
+ExecStart=/usr/local/sbin/ovs-dpdk.sh
+TimeoutStartSec=0
+
+[Install]
+WantedBy=default.target
+
+```
+
+
+``` /usr/local/sbin/ovs-dpdk.sh
+#!/bin/bash
+
+modprobe vfio-pci
+dpdk-devbind -u <pci_addr>
+dpdk-devbind -b vfio-pci <pci_addr>
+
+ovs-vsctl set Open_vSwitch . other_config:pmd-cpu-mask=6
+ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-socket-mem="1024,0"
+ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-init=true
+```
+
+Enable the new service:
+
+```
+# systemctl enable ovs-dpdk.service
+```
+
+
+## OVS-DPDK in oVirt
+ 
+* Upon creating a new cluster, choose ovs as the switch type.
+* When configuring host network interfaces, DPDK-bound NICs will be presented in the system.
+* When creating a network on DPDK device, an OVS bridge with datapath_type=’netdev’ will be automatically attached to dpdk device will be created.
+
+[ovs-vsctl]: (/images/ovs-dpdk.png)
+ 
+### VMs
+ 
+VMs will be created with vhostuser network interface.
+ 
+Vhostuser XML file:
+
+``` 
+<interface type="vhostuser">
+            <address bus="0x00" domain="0x0000" function="0x0" slot="0x04" type="pci" />
+            <mac address="00:1a:4a:16:01:54" />
+            <model type="virtio" />
+            <source mode="server" path="/var/run/vdsm/vhostuser/ee844c42-7075-3bf5-86c2-ec831f287c22" type="unix" />
+            <filterref filter="vdsm-no-mac-spoofing" />
+            <link state="up" />
+            <bandwidth />
+        </interface>
+```
+### DPDK in the guest
+
+DPDK can be used in the guest in order to run a high speed packet forwarding program between vhostuser ports.
+DPDK package and the desired application should be installed on the guest and hugepages should be configured.
+
+## Limitations
+* Shared huge pages are required for OVS-DPDK. If they are not shared, vhostuser interface will appear but there won’t be any network communication.
```

Comment:
how is this set? via the custom properties mentioned above? if so

" you must configure shared huge pages on each VM you create. ..."